crates/ty_python_semantic/resources/mdtest/generics/pep695/paramspec.md

```diff
@@ -62,3 +62,503 @@ Other values are invalid.
+
+```py
+from typing import Callable
+
+def foo[**P](c: Callable[P, int]) -> None:
+    def nested1(*args: P.args, **kwargs: P.kwargs) -> None: ...
+
+    # error: [invalid-type-form] "`P.kwargs` is valid only in `**kwargs` annotation: Did you mean `P.args`?"
+    # error: [invalid-type-form] "`P.args` is valid only in `*args` annotation: Did you mean `P.kwargs`?"
+    def nested2(*args: P.kwargs, **kwargs: P.args) -> None: ...
+
+    # TODO: error
+    def nested3(*args: P.args) -> None: ...
+
+    # TODO: error
+    def nested4(**kwargs: P.kwargs) -> None: ...
+
+    # TODO: error
+    def nested5(*args: P.args, x: int, **kwargs: P.kwargs) -> None: ...
+```
+
+And, they need to be used together.
+
+```py
+def foo[**P](c: Callable[P, int]) -> None:
+    # TODO: error
+    def nested1(*args: P.args) -> None: ...
+
+    # TODO: error
+    def nested2(**kwargs: P.kwargs) -> None: ...
+
+class Foo[**P]:
+    # TODO: error
+    args: P.args
+
+    # TODO: error
+    kwargs: P.kwargs
+```
+
+## Semantics of `P.args` and `P.kwargs`
+
+The type of `args` and `kwargs` inside the function is `P.args` and `P.kwargs` respectively instead
+of `tuple[P.args, ...]` and `dict[str, P.kwargs]`.
+
+### Passing `*args` and `**kwargs` to a callable
+
+```py
+from typing import Callable
+
+def f[**P](func: Callable[P, int]) -> Callable[P, None]:
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> None:
+        reveal_type(args)  # revealed: P@f.args
+        reveal_type(kwargs)  # revealed: P@f.kwargs
+        reveal_type(func(*args, **kwargs))  # revealed: int
+
+        # error: [invalid-argument-type] "Argument is incorrect: Expected `P@f.args`, found `P@f.kwargs`"
+        # error: [invalid-argument-type] "Argument is incorrect: Expected `P@f.kwargs`, found `P@f.args`"
+        reveal_type(func(*kwargs, **args))  # revealed: int
+
+        # error: [invalid-argument-type] "Argument is incorrect: Expected `P@f.args`, found `P@f.kwargs`"
+        reveal_type(func(args, kwargs))  # revealed: int
+
+        # Both parameters are required
+        # TODO: error
+        reveal_type(func())  # revealed: int
+        reveal_type(func(*args))  # revealed: int
+        reveal_type(func(**kwargs))  # revealed: int
+    return wrapper
+```
+
+### Operations on `P.args` and `P.kwargs`
+
+The type of `P.args` and `P.kwargs` behave like a `tuple` and `dict` respectively. Internally, they
+are represented as a type variable that has an upper bound of `tuple[object, ...]` and
+`Top[dict[str, Any]]` respectively.
+
+```py
+from typing import Callable, Any
+
+def f[**P](func: Callable[P, int], *args: P.args, **kwargs: P.kwargs) -> None:
+    reveal_type(args + ("extra",))  # revealed: tuple[object, ...]
+    reveal_type(args + (1, 2, 3))  # revealed: tuple[object, ...]
+    reveal_type(args[0])  # revealed: object
+
+    reveal_type("key" in kwargs)  # revealed: bool
+    # TODO: Should be `object | None`
+    reveal_type(kwargs.get("key"))  # revealed: object
```

Comment:
I don't think this TODO is accurate, because `None` is a subtype of `object`: `object | None` just immediately simplifies to `object`. The existing assertion is correct!

```suggestion
    reveal_type(kwargs.get("key"))  # revealed: object
```

Reply:
🤦 thanks! I was a little bit confused to see this but yes it's obvious now